modules/internal/ChapelArray.chpl

```diff
@@ -3211,6 +3211,68 @@ module ChapelArray {
       else
         a.add(e);
   }
+
+  //
+  // BaseSparseDom operator overloads
+  //
+  proc +=(ref sd: domain, inds: [] sd.idxType) where isSparseDom(sd) &&
+    sd.rank==1 {
+
+    if inds.size == 0 then return;
+
+    sd._value.dsiBulkAdd(inds);
+  }
+
+  proc +=(ref sd: domain, inds: [] sd.rank*sd.idxType) where isSparseDom(sd) &&
+    sd.rank>1 {
+
+    if inds.size == 0 then return;
+
+    sd._value.dsiBulkAdd(inds);
+  }
+
+  // Currently this is not optimized for addition of a sparse
+  // This assumes that a domain is sorted by row, which may not always be true.
+  // There could be domains sorted by column instead.
+  // It would be nice if there was a common interface in BaseDom to query
+  // whether a domain was row-major or column-major to distinguish these.
+  // The current implementation is a work-around for this shortcoming.
+  proc +=(ref sd: domain, d: domain)
+  where sd._value: DefaultSparseDom && d.rank==sd.rank && sd.idxType==d.idxType {
+
+    if d.size == 0 then return;
+
+    type _idxType = if sd.rank==1 then int else sd.rank*int;
+    const indCount = d.numIndices;
+    const arr: [{0..#indCount}] _idxType;
+
+    //this could be a parallel loop. but ranks don't match -- doesn't compile
+    for (a,i) in zip(arr,d) do a=i;
+
+    sd._value.dsiBulkAdd(arr, true, true, false);
```

Comment:
If I understand correctly, dsiBulkAdd() says whether or not the `arr` value being passed in is sorted.  But here, if the RHS domain is a CSC, they won't be if I'm understanding your PR's message properly.  So I would've expected this to pass in `false` for `sorted` (in which case I'd also wonder if the where clause restricting the LHS to be default sparse could also be removed safely).